ruby/computer_science/space_complexity.md

```diff
@@ -0,0 +1,141 @@
+
+That's why we won't be diving into examples for other Big O notations with space complexity. We'd have to come up with convuluted examples that didn't represent most code you'll write. If you do come across a good real world example in your own code then do let us know and we may consider adding it here for others to consider.
+
+#### Other considerations
+
+One of the common areas that causes confusion when considering space complexity is what constitutes using space in the context of an algorithm. In an earlier example we wrote methods that duplicated an array and hash argument. We did that to be explicit. But what if we'd written the method as:
+
+~~~ruby
+def sum_arr(arr)
+  sum = 0
+  arr.each do |number|
+    sum += number
+  end
+  return sum
+end
+~~~
+
+When a data structure is passed in as the argument, especially for languages that pass arrays by reference rather than value, it can be a bit unclear if that method considers the space used by that data structure when calculating its space complexity. If you didn't count it, then it would be easy for all our methods to have great space usage on paper because we put the onus on the caller to allocate that space. If we did count it, but the data structure was created for use by many different methods then is the space complexity for all those methods O(N) when they aren't utilising additional space. Then consider that if your method receives an array as an input and loops it, then an index must be created for the loop which uses additional space.
+
+The first answer [to this question](https://cs.stackexchange.com/questions/127933/analyzing-space-complexity-of-passing-data-to-function-by-reference) provides some great context to the question and gives some thought provoking answers.
+
+Ultimately when you consider Big O measures the worst case scenario it would be easier to err on the side of caution and do consider the space of arguments passed to your method.
+
+### Wrapping up
+
+Measuring the complexity of your algorithms, whether time or space, can be difficult. It takes practice and consideration. For most practice code you write it's not something that will cross your mind, especially as you wrestle with getting your code to work.
+
+Once your code is working though, and you might be looking to refactor it, it's definitely worth taking a moment to consider if the code is as efficient as it could be. Are you creating unecessary variables? Or does your algorithm use a data structure with a worse time complexity for what it's mostly used for than another data structure would have been.
```

Comment:
```suggestion
Once your code is working though, and you might be looking to refactor it, it's definitely worth taking a moment to consider if the code is as efficient as it could be. Are you creating unnecessary variables? Or does your algorithm use a data structure with a worse time complexity for what it's mostly used for than another data structure would have been.
```